lib/Backends/CPU/FunctionSpecializer.cpp

```diff
@@ -343,7 +339,34 @@ class FunctionSpecializer {
 } // namespace
 
 void LLVMIRGen::performSpecialization() {
-  FunctionSpecializer FuncSpecializer(llmodule_->getFunction("main"));
+  // Specialize all invocations of main functions with constant parameters to
+  // produce a better optimized code.
+  auto *func = getModule().getFunction("jitmain");
+  while (func) {
+    // Call of a "main" function is always the last instruction before return.
+    auto instrBeforeTerminator =
+        std::prev(func->back().getTerminator()->getIterator());
+    auto *call = dyn_cast<llvm::CallInst>(instrBeforeTerminator);
+    if (!call)
+      break;
+    // Specialize only calls of all "main" functions.
+    func = call->getCalledFunction();
+    if (!func || !func->getName().startswith("main"))
+      break;
+    call = specializeCallWithConstantArguments(call);
+    func = call ? call->getCalledFunction() : nullptr;
+  }
+
+  // Specialize calls in all main functions.
+  llvm::SmallVector<llvm::Function *, 32> mainFunctions;
+  for (auto &F : getModule()) {
+    if (!F.getName().startswith("main")) {
```

Comment:
Not a huge fan of the hardcoded "main". At least, could it be hold in some global/whatever so that it is not repeated across the file.